api/controller.py

```diff
@@ -619,7 +651,69 @@ def borrow(self, data_source, identifier_type, identifier, mechanism_id=None):
         headers = { "Content-Type" : OPDSFeed.ACQUISITION_FEED_TYPE }
         return Response(content, status_code, headers)
 
-    def fulfill(self, data_source, identifier_type, identifier, mechanism_id=None, do_get=None):
+    def best_lendable_pool(self, library, patron, identifier_type, identifier,
+                           mechanism_id):
+        """Of the available LicensePools for the given Identifier, return the
+        one that's the best candidate for loaning out right now.
+        """
+        # Turn source + identifier into a set of LicensePools
+        pools = self.load_licensepools(
+            self.library, identifier_type, identifier
+        )
+        if isinstance(pools, ProblemDetail):
+            # Something went wrong.
+            return pools
+
+        best = None
+        mechanism = None
+        problem_doc = None
+
+        existing_loans = self._db.query(Loan).filter(
+            Loan.license_pool_id.in_([lp.id for lp in pools]),
+            Loan.patron==patron
+        ).all()
+        if existing_loans:
+            return ALREADY_CHECKED_OUT
+
+        # We found a number of LicensePools. Try to locate one that
+        # we can actually loan to the patron.
+        for pool in pools:
+            problem_doc = self.apply_borrowing_policy(patron, pool)
+            if problem_doc:
+                # As a matter of policy, the patron is not allowed to borrow
+                # this book.
+                continue
+
+            # Beyond this point we know that site policy does not prohibit
+            # us from lending this pool to this patron.
+
+            if mechanism_id:
+                # But the patron has requested a license pool that
+                # supports a specific delivery mechanism. This pool
+                # must offer that mechanism.
+                mechanism = self.load_licensepooldelivery(pool, mechanism_id)
+                if isinstance(mechanism, ProblemDetail):
+                    problem_doc = mechanism
+                    continue
+
+            # Beyond this point we have a license pool that we can
+            # actually loan or put on hold.
+
+            # But there might be many such LicensePools, and we want
+            # to pick the one that will get the book to the patron
+            # with the shortest wait.
+            if (not best
+                or pool.licenses_available > best.licenses_available
+                or pool.patrons_in_hold_queue < best.patrons_in_hold_queue):
```

Comment:
This could also take into account the total licenses owned. A pool with 10 holds but 10 licenses owned is better that a pool with 9 holds but 1 license owned.